diff --git a/Makefile b/Makefile
old mode 100644
new mode 100755
index b7fb22d..5e1e62b
--- a/Makefile
+++ b/Makefile
@@ -360,6 +360,8 @@ grade:
 ##
 ## FOR submissions
 ##
+diff:
+	git diff origin/fs HEAD > commit.patch
 
 submit-check:
 	@if ! test -d .git; then \
diff --git a/user/ls.c b/user/ls.c
old mode 100644
new mode 100755
index 39ab074..7c46eb8
--- a/user/ls.c
+++ b/user/ls.c
@@ -3,7 +3,7 @@
 #include "user/user.h"
 #include "kernel/fs.h"
 #include "kernel/fcntl.h"
-
+aaaaa
 char*
 fmtname(char *path)
 {
